[Merged by Bors] - Empty Set and Failures support for Hare consensus

sudachen opened the pull request:
## Motivation
Closes #1967
Closes #2311
Closes #2404

## Changes
Hare always finished with a result

If preround fails empty set is used for hare continue
If hare fails empty set is considered as hare result

sudachen committed:
added empty set support

sudachen committed:
added empty set support

sudachen committed:
remove unnecesarry panic

sudachen committed:
remove unnecesarry panic

sudachen committed:
readded hare failures test

sudachen committed:
readded hare failures test

sudachen committed:
added error information

sudachen committed:
added error information

sudachen committed:
remove unnecesarry panic

sudachen committed:
remove unnecesarry panic

sudachen commented:
bors try

sudachen commented:
bors try

sudachen commented:
bors try

sudachen commented:
bors try

noamnelke committed:
Only report receiving a block once (#2437)

## Motivation
We've seen cases in system tests where nodes report getting several multiples of the number of blocks that are supposed to be in a layer. It appears to happen because we log receiving a block before checking if we know it already, at which point block processing stops.

## Changes
Log `got new block` only after checking for duplicates.

## Test Plan
Since this only affects logging, no new tests are needed.

## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

Commit:
p2p: use the same shutdown context for all components (#2433)

## Motivation
<!-- Please mention the issue fixed by this PR or detailed motivation -->
Closes #1313 
<!-- `Closes #XXXX, closes #XXXX, ...` links mentioned issues to this PR and automatically closes them when this it's merged -->

## Changes
<!-- Please describe in detail the changes made -->
make all p2p components share the same shutdown context so they can be in sync wrt to shutdown timeing.

## Test Plan
<!-- Please specify how these changes were tested 
(e.g. unit tests, manual testing, etc.) -->
unit tests

## TODO
<!-- This section should be removed when all items are complete -->
- [ ] Explain motivation or link existing issue(s)
- [ ] Test changes and document test plan
- [ ] Update documentation as needed

## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

noamnelke commented on hare/messagevalidation.go:
```diff
@@ -238,21 +238,23 @@ func (v *syntaxContextValidator) SyntacticallyValidateMessage(ctx context.Contex
+				log.String("sender_id", m.PubKey.ShortString()),
+				types.LayerID(m.InnerMsg.InstanceID),
+				log.String("msg_type", m.InnerMsg.Type.String()))
+			return false
+		}
 
-	if len(m.InnerMsg.Values) == 0 {
-		logger.With().Warning("syntax validation failed: set is empty",
-			log.String("sender_id", m.PubKey.ShortString()),
-			types.LayerID(m.InnerMsg.InstanceID),
-			log.String("msg_type", m.InnerMsg.Type.String()))
-		return false
-	}
+		if len(m.InnerMsg.Values) == 0 {
+			logger.With().Warning("syntax validation failed: set is empty",
+				log.String("sender_id", m.PubKey.ShortString()),
+				types.LayerID(m.InnerMsg.InstanceID),
+				log.String("msg_type", m.InnerMsg.Type.String()))
+			return false
+		}*/
```
Let's remove the commented-out code.

noamnelke commented on hare/hare.go:
```diff
@@ -160,21 +160,23 @@ var ErrTooLate = errors.New("consensus process finished too late")
-		i++
-	}
+	var blocks []types.BlockID
+	if output.Completed() {
```
This seems redundant to me.

If `output.Completed() == false` then `len(output.Set()) == 0` and then the loop runs zero times and with an empty blocks list `h.validate(blocks)` does nothing and returns `true`.

noamnelke commented on hare/hare.go:
```diff
@@ -241,8 +255,9 @@ func (h *Hare) onTick(ctx context.Context, id types.LayerID) {
-		logger.With().Error("no blocks for consensus", log.Err(err))
-		return
+		// we have to try achieve consensus on empty set or fail, otherwise it will be the delayed consensus process forever
+		logger.With().Error("", log.Err(err))
```
Let's add a message here. Maybe `no blocks found for hare, using empty set`?

noamnelke commented on hare/hare.go:
```diff
@@ -254,12 +269,13 @@ func (h *Hare) onTick(ctx context.Context, id types.LayerID) {
 	if err != nil {
-		logger.With().Warning("could not register consensus process on broker", log.Err(err))
+		logger.With().Error("could not register consensus process on broker", log.Err(err))
+		err = errors.New("could not register consensus process on broker")
```
What's the point of changing the error here? We do nothing with it other than check for nilness.

noamnelke commented on hare/hare.go:
```diff
@@ -278,6 +295,9 @@ var (
 // Returns error iff the request for the upper is too old.
 func (h *Hare) GetResult(lid types.LayerID) ([]types.BlockID, error) {
+
+	h.With().Warning("GetResult", lid)
```
I assume this is a warning for debugging purposes - let's change to info/debug (feels like a debug to me, but up to you).

noamnelke commented on hare/hare.go:
```diff
@@ -297,12 +317,9 @@ func (h *Hare) outputCollectionLoop(ctx context.Context) {
+				h.WithContext(ctx).With().Warning("error collecting output from hare", log.Err(err))
 			}
-
 			// either way, unregister from broker
```
This comment is no longer relevant.

noamnelke commented on hare/algorithm_test.go:
```diff
@@ -186,13 +186,15 @@ func TestConsensusProcess_Start(t *testing.T) {
 	proc.SetInbox(inbox)
-	err := proc.Start(context.TODO())
-	assert.Equal(t, "instance started with an empty set", err.Error())
+	/*
+		proc.s = NewDefaultEmptySet()
+		err := proc.Start(context.TODO())
+		assert.Equal(t, "instance started with an empty set", err.Error())
+	*/
```
Remove comment.

noamnelke commented on hare/messagevalidation_test.go:
```diff
@@ -166,7 +168,7 @@ func TestMessageValidator_Aggregated(t *testing.T) {
 	inner := msgs[0].InnerMsg.Values
 	msgs[0].InnerMsg.Values = nil
-	r.Equal(errInnerSyntax, validator.validateAggregatedMessage(context.TODO(), agg, funcs))
+	//r.Equal(errInnerSyntax, validator.validateAggregatedMessage(context.TODO(), agg, funcs))
```
Remove comment.

noamnelke commented on hare/xfailure_test.go:
```diff
@@ -0,0 +1,332 @@
+package hare
```
The name of this file is not great. It tests both failures and successes.

At first I thought you missed writing success tests as I reviewed this file last. It might be confusing for someone looking for a test in the future.

noamnelke reviewed (approved):
Approved. Had just minor comments that should be fixed before merging.

lrettig commented on activation/activation.go:
```diff
@@ -386,22 +386,32 @@ func (b *Builder) PublishActivationTx(ctx context.Context) error {
 		}
 	}
+
+	b.log.With().Info("chellenge ready")
```
```suggestion
	b.log.With().Info("challenge ready")
```

lrettig commented on hare/hare.go:
```diff
@@ -206,16 +209,19 @@ func (h *Hare) onTick(ctx context.Context, id types.LayerID) {
 	}
 
+	defer func() {
+		// it must don't return without starting consensus process or mark result as fail
```
```suggestion
		// it must not return without starting consensus process or mark result as fail
```

lrettig commented on hare/hare.go:
```diff
@@ -241,8 +255,9 @@ func (h *Hare) onTick(ctx context.Context, id types.LayerID) {
-		return
+		// we have to try achieve consensus on empty set or fail, otherwise it will be the delayed consensus process forever
+		logger.With().Error("", log.Err(err))
+		//return
```
Are you sure you don't want to return here?

sudachen replied:
Not sure. But If there is no difference, why not?

lrettig commented on hare/hare.go:
```diff
@@ -241,8 +255,9 @@ func (h *Hare) onTick(ctx context.Context, id types.LayerID) {
 	if err != nil {
-		logger.With().Error("no blocks for consensus", log.Err(err))
-		return
+		// we have to try achieve consensus on empty set or fail, otherwise it will be the delayed consensus process forever
```
I don't understand this comment. Why would it "otherwise delay the consensus process forever"? If hare never runs for layer N, it will still try to run for layer N+1

sudachen replied:
It was before I added defer finishing hare with "notCompleted" state. So it was leaving this hare run incomplete for the tortoise.

sudachen committed:
fixed problems found during the review

sudachen committed:
fixed problems found during the review

sudachen committed:
run hare even node does not see blockc in the layer

sudachen committed:
run hare even node does not see blockc in the layer

sudachen commented:
bors try

sudachen commented:
bors try

AmitShaul committed:
moved teardown process to GCP using 'google cloud tasks' (#2442)

## Motivation
making tests finish faster

## Changes
moved automation teardown to GCP using 'google cloud tasks'
this includes the following tasks to be executed on the cloud:
* remove node pool
* delete fluent bits clusterrole binding object
* dumping ES logs to main ES if necessary
* removing namespace

added app engine code under automation 'tests' directory

on the node side:
changed 'cached_epoch_id' from string to int where -1 replaced '(none)'

sudachen committed:
Merge branch 'empty_set_support' of github.com:spacemeshos/go-spacemesh into empty_set_support

sudachen commented:
bors try

sudachen commented:
bors try

AmitShaul commented:
bors try

sudachen commented:
bors try

sudachen commented:
bors try

sudachen commented:
bors try

sudachen commented:
bors merge

sudachen commented:
bors merge
